hacep/src/main/java/org/kie/hacep/consumer/DroolsConsumerHandler.java

```diff
@@ -44,45 +45,85 @@
     private KieSessionContext kieSessionContext;
     private CommandHandler commandHandler;
-    private SnapshotInfos infos;
+    private SnapshotInfos snapshotInfos;
     private boolean shutdown;
 
     public DroolsConsumerHandler(Producer producer, EnvConfig envConfig) {
-        this.config = envConfig;
-        this.snapshooter = new DefaultSessionSnapShooter(config);
-        initializeKieSessionFromSnapshot(config);
+        this.envConfig = envConfig;
+        this.sessionSnapShooter = new DefaultSessionSnapShooter(this.envConfig);
+        initializeKieSessionContext(this.envConfig);
         this.producer = producer;
-        commandHandler = new CommandHandler(kieSessionContext, config, producer, snapshooter);
-        if (config.isUnderTest()) {
+        this.commandHandler = new CommandHandler(this.kieSessionContext, this.envConfig, producer, this.sessionSnapShooter);
+        if (this.envConfig.isUnderTest()) {
             loggerForTest = PrinterUtil.getKafkaLoggerForTest(envConfig);
         }
     }
 
-    private void initializeKieSessionFromSnapshot(EnvConfig config) {
-        if(config.isSkipOnDemanSnapshot()) {// if true we reads the snapshots and wait until the first leaderElectionUpdate
-            this.infos = snapshooter.deserialize();
-            this.kieSessionContext = createSessionHolder(infos);
+    private void initializeKieSessionContext(EnvConfig config) {
+        if(config.isSkipOnDemandSnapshot()) {// if true we reads the snapshots and wait until the first leaderElectionUpdate
+            initializeSessionContextWithSnapshotCheck();
         } else{
-            kieSessionContext = new KieSessionContext();
-            createClasspathSession( kieSessionContext );
+            createAndInitializeSessionContextWithoutSnapshot();
+        }
+    }
+
+    private void initializeSessionContextWithSnapshotCheck() {
+        this.snapshotInfos = this.sessionSnapShooter.deserialize();
+        /* if the Snapshot is ok we use the KieContainer and KieSession from the infos,
+           otherwise (empty system) we create kiecontainer and kieSession from the envConfig's GAV */
+        if (this.snapshotInfos != null) {
+            initializeSessionContextFromSnapshot();
+        }else{
+            createAndInitializeSessionContextWithoutSnapshot();
+        }
+    }
+
+    private void createAndInitializeSessionContextWithoutSnapshot() {
+        KieServices srv = KieServices.get();
+        if (srv != null) {
+            KieContainer kieContainer;
+            if(envConfig.isUpdatableKJar()){
+                if (logger.isInfoEnabled()) {logger.info("Creating new KieContainer with KJar:{}", envConfig.getKjarGAV());}
+                String parts[] = envConfig.getKjarGAV().split(":");
+                kieContainer = srv.newKieContainer(srv.newReleaseId(parts[0], parts[1], parts[2]));
+                KieScanner scanner = srv.newKieScanner(kieContainer);
+                scanner.scanNow();
```

Comment:
What does it mean? Is it a blocking call to initialize the `KieContainer`? Is it needed even during an update?

Reply:
Is a scan if the snapshot isn't available during the start, we want to drop this and enable only via snapshot and the update at runtime ?